src/StackExchange.Redis/ConnectionMultiplexer.cs

```diff
@@ -2214,30 +2350,61 @@ public ConnectionMultiplexer GetSentinelMasterConnection(ConfigurationOptions co
                     return sentinelConnectionChildren[config.ServiceName];
             }
 
-            // Get an initial endpoint - try twice
-            EndPoint newMasterEndPoint = GetConfiguredMasterForService(config.ServiceName)
-                                      ?? GetConfiguredMasterForService(config.ServiceName);
+            int attempts = 0;
+            bool success = false;
+            ConnectionMultiplexer connection = null;
 
-            if (newMasterEndPoint == null)
+            do
             {
-                throw new RedisConnectionException(ConnectionFailureType.UnableToConnect,
-                    $"Sentinel: Failed connecting to configured master for service: {config.ServiceName}");
-            }
+                attempts++;
 
-            // Replace the master endpoint, if we found another one
-            // If not, assume the last state is the best we have and minimize the race
-            if (config.EndPoints.Count == 1)
-            {
-                config.EndPoints[0] = newMasterEndPoint;
-            }
-            else
+                // Get an initial endpoint - try twice
+                EndPoint newMasterEndPoint = GetConfiguredMasterForService(config.ServiceName)
+                                             ?? GetConfiguredMasterForService(config.ServiceName);
+
+                if (newMasterEndPoint == null)
+                {
+                    throw new RedisConnectionException(ConnectionFailureType.UnableToConnect,
+                        $"Sentinel: Failed connecting to configured master for service: {config.ServiceName}");
+                }
+
+                EndPoint[] slaveEndPoints = GetSlavesForService(config.ServiceName)
+                                            ?? GetSlavesForService(config.ServiceName);
+
+                // Replace the master endpoint, if we found another one
+                // If not, assume the last state is the best we have and minimize the race
+                if (config.EndPoints.Count == 1)
+                {
+                    config.EndPoints[0] = newMasterEndPoint;
+                }
+                else
+                {
+                    config.EndPoints.Clear();
+                    config.EndPoints.TryAdd(newMasterEndPoint);
+                }
+
+                foreach (var slaveEndPoint in slaveEndPoints)
+                    config.EndPoints.TryAdd(slaveEndPoint);
+
+                connection = ConnectImpl(config, log);
+
+                // verify role is master according to:
+                // https://redis.io/topics/sentinel-clients
+                if (connection.GetServer(newMasterEndPoint).Role() == RedisLiterals.master)
```

Comment:
looking at the `ROLE` command on redis, it is *much* more complex and nuanced than the shape of the command you're throwing onto `IServer`; can I suggest *don't* add this to the public API right now - this needs a lot more consideration, and it is unrelated to this PR; you might want to just use `Execute`, which can invoke any command, and get the result back out from there

Reply:
example usage:

``` c#
var role = (string)((RedisResult[])server.Execute("ROLE"))[0];
```